scala-package/core/src/main/scala/ml/dmlc/mxnet/module/Module.scala

```diff
@@ -437,8 +442,104 @@ class Module(symbolVar: Symbol,
+    } else {
+      updater.foreach {
+        case cachedStates: MXKVStoreCachedStates =>
+          val target = new BufferedOutputStream(new FileOutputStream(fname))
+          try {
+            target.write(cachedStates.serializeState())
+          } finally {
+            target.close()
+          }
+        case _ =>
+          logger.warn("Updater does not have states, skip saving to {}", fname)
+      }
+    }
+  }
+
+  /**
+   * Load optimizer (updater) state from file
+   * @param fname Path to input states file.
+   */
+  def loadOptimizerStates(fname: String): Unit = {
+    require(optimizerInitialized, "Optimizer should be initialized before loading.")
+    if (updateOnKVStore) {
+      kvstore.foreach(_.loadOptimizerStates(fname))
+    } else {
+      updater.foreach {
+        case cachedStates: MXKVStoreCachedStates =>
+          val bis = new BufferedInputStream(new FileInputStream(fname))
+          try {
+            val bArray = Stream.continually(bis.read).takeWhile(-1 !=).map(_.toByte).toArray
+            cachedStates.deserializeState(bArray)
+          } finally {
+            bis.close()
+          }
+        case _ =>
+          logger.warn("Updater does not have states, skip loading from {}", fname)
+      }
+    }
+  }
+
+  /**
+   * Save current progress to checkpoint.
+   * Use mx.callback.module_checkpoint as epoch_end_callback to save during training.
+   * @param prefix The file prefix to checkpoint to
+   * @param epoch The current epoch number
+   * @param saveOptStates Whether to save optimizer states for continue training
+   */
+  def saveCheckpoint(prefix: String, epoch: Int, saveOptStates: Boolean = false): Unit = {
+    symbol.save(s"$prefix-symbol.json")
+    val paramName = "%s-%04d.params".format(prefix, epoch)
+    saveParams(paramName)
+    logger.info("Saved checkpoint to {}", paramName)
+    if (saveOptStates) {
+      val stateName = "%s-%04d.states".format(prefix, epoch)
+      saveOptimizerStates(stateName)
+      logger.info("Saved optimizer state to {}", stateName)
+    }
+  }
 }
 
 object Module {
-  // TODO: Load & save model
+  /**
+   * Create a model from previously saved checkpoint.
+   * @param prefix Path prefix of saved model files. You should have "prefix-symbol.json",
+   *               "prefix-xxxx.params", and optionally "prefix-xxxx.states",
+   *               where xxxx is the epoch number.
+   * @param epoch Epoch to load.
+   * @param loadOptimizerStates Whether to load optimizer states.
+   *                            Checkpoint needs to have been made with saveOptimizerStates=True
+   * @param dataNames Input data names.
+   * @param labelNames Input label names
+   * @param contexts Default is cpu().
+   * @param workLoadList  Default `None`, indicating uniform workload.
+   * @param fixedParamNames Default `None`, indicating no network parameters are fixed.
+   */
+  def loadCheckpoint(prefix: String, epoch: Int, loadOptimizerStates: Boolean = false,
```

Comment:
About this part. I'm thinking about making a better save/load format across languages. Something like keras's model format